opm/material/densead/EvaluationV.hpp

```diff
@@ -0,0 +1,602 @@
+// -*- mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*-
+// vi: set et ts=4 sw=4 sts=4:
+/*
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 2 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+
+  Consult the COPYING file in the top-level source directory of this
+  module for the precise wording of the license and the list of
+  copyright holders.
+*/
+/*!
+ * \file
+ *
+ * \brief Representation of an evaluation of a function and its derivatives w.r.t. a set
+ *        of variables in the localized OPM automatic differentiation (AD) framework.
+ */
+#ifndef OPM_DENSEAD_EVALUATION_V_HPP
+#define OPM_DENSEAD_EVALUATION_V_HPP
+
+#include "MathV.hpp"
+
+#include <opm/material/common/Valgrind.hpp>
+
+#include <dune/common/version.hh>
+
+#include <vector>
+#include <cmath>
+#include <cassert>
+#include <cstring>
+#include <iostream>
+#include <algorithm>
+
+namespace Opm {
+namespace DenseAd {
+
+/*!
+ * \brief Represents a function evaluation and its derivatives w.r.t. a fixed set of
+ *        variables.
+ */
+template <class ValueT>
+class EvaluationV
```

Comment:
it is better to generate a specialization for `Evaluation` with `-1` as the number of parameters IMO. confer https://github.com/OPM/opm-material/blob/master/bin/genEvalSpecializations.py .

Reply:
I am not sure about this.  What does physically mean an `Evaluation` with `-1` number of the variables?

Reply:
> What does physically mean an Evaluation with -1 number of the variables?

That's our choice to decide. I propose "number of derivatives is determined at runtime".

Reply:
if you don't like this to be a raw number, a constant can be added, e.g.:

```c++
typedef Opm::DenseAd::Evaluation<double, Opm::DenseAd::runtimeDetermined> EvaluationV;
EvaluationV x(/*numDeriv=*/2);
```

Reply:
> That's our choice to decide. I propose "number of derivatives is determined at runtime".

Okay. You are suggesting to use specialization of `Evaluation` with -1 number of variables instead of a new `EvaluationV`.  I do not have problem with that, as long as it will work for my purpose.